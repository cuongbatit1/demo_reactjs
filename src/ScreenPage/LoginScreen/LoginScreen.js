import * as React from "react";
import { memo, useCallback } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Link, useNavigate } from "react-router-dom";
import { useTranslation } from "react-i18next";

// material-ui
import { useTheme } from "@mui/material/styles";
import { Divider, Grid, Typography, useMediaQuery } from "@mui/material";
import Button from "@mui/material/Button";
import Box from "@mui/material/Box";
import Checkbox from "@mui/material/Checkbox";
import FormControl from "@mui/material/FormControl";
import FormHelperText from "@mui/material/FormHelperText";
import InputAdornment from "@mui/material/InputAdornment";
import Visibility from "@mui/icons-material/Visibility";
import VisibilityOff from "@mui/icons-material/VisibilityOff";
import InputLabel from "@mui/material/InputLabel";
import OutlinedInput from "@mui/material/OutlinedInput";
import IconButton from "@mui/material/IconButton";
import TextField from "@mui/material/TextField";

import AuthWrapper1 from "../../component/AuthWrapper1";
import AuthCardWrapper from "../../component/AuthCardWrapper";
import {
  changeEmailLoginAction,
  changePasswordLoginAction,
  handleCheckRememberMe,
  handleShowHidePasswordLogin,
  onBlurEmailLoginAction,
  onBlurPasswordLoginAction,
  submitLogin,
} from "../../redux/actions/loginAction";
import DialogMessage from "../Dialog/DialogMessage";
import LoadingModalDialog from "../Dialog/LoadingModal";

const EmailInputComponent = () => {
  const dispatch = useDispatch();
  const { t } = useTranslation();
  // ---START--- DATA FROM REDUX
  const email = useSelector((state) => state.loginReducer.email);
  const errorEmail = useSelector((state) => state.loginReducer.errorEmail);
  // ---END--- DATA FROM REDUX

  // --- START --- funtion handle in component useCallback

  const onChangeEmail = useCallback((event) => dispatch(changeEmailLoginAction(event)), [dispatch]);

  const onBlurEmail = useCallback(() => dispatch(onBlurEmailLoginAction()), [dispatch]);

  // --- END --- funtion handle in component
  return (
    <>
      {console.log("EmailInputComponent Render")}
      <TextField
        sx={{ mb: 2 }}
        error={errorEmail !== "" ? true : false}
        // defaultValue={email}
        value={email}
        label={t("email")}
        variant="outlined"
        style={{ width: "100%" }}
        type="email"
        onChange={onChangeEmail}
        onBlur={onBlurEmail}
        helperText={errorEmail}
      />
    </>
  );
};

const PasswordInputComponent = () => {
  const dispatch = useDispatch();
  const { t } = useTranslation();
  // ---START--- DATA FROM REDUX
  const password = useSelector((state) => state.loginReducer.password);
  const showPassword = useSelector((state) => state.loginReducer.showPassword);
  const errorPassword = useSelector((state) => state.loginReducer.errorPassword);

  // ---END--- DATA FROM REDUX

  // --- START --- funtion handle in component useCallback

  const onChangePassword = useCallback(
    (event) => dispatch(changePasswordLoginAction(event)),
    [dispatch]
  );

  const onBlurPassword = useCallback(() => dispatch(onBlurPasswordLoginAction()), [dispatch]);

  const showHidePassword = useCallback(() => dispatch(handleShowHidePasswordLogin()), [dispatch]);
  // --- END --- funtion handle in component
  return (
    <>
      {console.log("PasswordInputComponent Render")}
      <FormControl
        error={errorPassword !== "" ? true : false}
        variant="outlined"
        sx={{ minWidth: 120, width: "100%" }}
      >
        <InputLabel htmlFor="outlined-adornment-password">{t("password")}</InputLabel>
        <OutlinedInput
          id="outlined-adornment-password"
          label={t("password")}
          type={showPassword ? "text" : "password"}
          // defaultValue={password}
          value={password}
          onChange={onChangePassword}
          onBlur={onBlurPassword}
          endAdornment={
            <InputAdornment position="end">
              <IconButton
                aria-label="toggle password visibility"
                onClick={showHidePassword}
                onMouseDown={(event) => event.preventDefault}
              >
                {showPassword ? <VisibilityOff /> : <Visibility />}
              </IconButton>
            </InputAdornment>
          }
        />
        <FormHelperText>{errorPassword}</FormHelperText>
      </FormControl>
    </>
  );
};

const CheckBoxActiveComponent = () => {
  const dispatch = useDispatch();
  // ---START--- DATA FROM REDUX
  const isRemember = useSelector((state) => state.loginReducer.isRemember);
  // ---END--- DATA FROM REDUX

  // --- START --- funtion handle in component useCallback

  const clickCheckBox = useCallback(() => dispatch(handleCheckRememberMe()), [dispatch]);
  // --- END --- funtion handle in component
  return (
    <>
      {console.log("CheckBoxActiveComponent Render")}
      <Box style={{}}>
        <Checkbox
          checked={isRemember}
          style={{ color: "black" }}
          inputProps={{
            "aria-label": "select all desserts",
          }}
          onChange={clickCheckBox}
        />
      </Box>
    </>
  );
};

function LoginScreen() {
  const theme = useTheme();
  const matchDownSM = useMediaQuery(theme.breakpoints.down("md"));
  const navigate = useNavigate();
  const { t } = useTranslation();
  const dispatch = useDispatch();
  // --- START --- funtion handle in component useCallback

  const onSubmitLogin = useCallback(() => dispatch(submitLogin(navigate)), [dispatch, navigate]);
  // --- END --- funtion handle in component
  return (
    <>
      <AuthWrapper1>
        <Grid container direction="column" justifyContent="flex-end" sx={{ minHeight: "100vh" }}>
          <Grid item xs={12}>
            <Grid
              container
              justifyContent="center"
              alignItems="center"
              sx={{ minHeight: "calc(100vh - 68px)" }}
            >
              <Grid item sx={{ m: { xs: 1, sm: 3 }, mb: 0 }}>
                <AuthCardWrapper>
                  <Grid container spacing={2} alignItems="center" justifyContent="center">
                    <Grid item xs={12}>
                      <Typography
                        variant="subtitle1"
                        sx={{
                          textDecoration: "none",
                          mb: 2,
                          textAlign: "center",
                        }}
                      >
                        Login with email
                      </Typography>
                      <EmailInputComponent />
                      <PasswordInputComponent />
                      <Grid
                        sx={{ mt: 2, mb: 2 }}
                        container
                        direction="row"
                        justifyContent="flex-start"
                        alignItems="center"
                      >
                        <CheckBoxActiveComponent />
                        <Typography sx={{ flexGrow: 1 }} variant="h6" component="h2">
                          {t("remember_me")}
                        </Typography>
                      </Grid>
                      <Button
                        sx={{ width: "100%", fontSize: 18 }}
                        variant="contained"
                        onClick={onSubmitLogin}
                      >
                        {t("sign_in")}
                      </Button>
                    </Grid>
                    <Grid item xs={12}>
                      <Divider />
                    </Grid>
                    <Grid item xs={12}>
                      <Grid item container direction="column" alignItems="center" xs={12}>
                        <Typography
                          component={Link}
                          to="/register/"
                          variant="subtitle1"
                          sx={{ textDecoration: "none" }}
                        >
                          Don&apos;t have an account?
                        </Typography>

                        <Typography
                          component={Link}
                          to="/about/"
                          variant="subtitle1"
                          sx={{ textDecoration: "none" }}
                        >
                          Go to About
                        </Typography>
                      </Grid>
                    </Grid>
                  </Grid>
                </AuthCardWrapper>
              </Grid>
            </Grid>
          </Grid>
          <Grid item xs={12} sx={{ m: 3, mt: 1 }}></Grid>
        </Grid>
      </AuthWrapper1>
      <DialogMessage />
      <LoadingModalDialog />
    </>
  );
}

export default LoginScreen;
