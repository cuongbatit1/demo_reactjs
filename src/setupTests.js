// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom";
import "@testing-library/react";
import { Provider } from "react-redux";
import { render as rtlRender } from "@testing-library/react";
import { REDUX_TOOL_KIT } from "./constants/AppConstants";
import configureStore from "./redux/configureStore";
// import configureStoreToolkit from "./reduxtoolkit/configureStoreToolkit";
import { MaterialUIControllerProvider } from "./context/MaterialUIControllerProvider";
import i18n from "./i18n";
import { BrowserRouter } from "react-router-dom";

function render(
  ui,
  {
    preloadedState,
    store = configureStore(preloadedState),
    // store = REDUX_TOOL_KIT ? configureStoreToolkit() : configureStore(),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    i18n.changeLanguage("en");
    return (
      <Provider store={store}>
        <MaterialUIControllerProvider>
          <BrowserRouter>{children}</BrowserRouter>
        </MaterialUIControllerProvider>
      </Provider>
    );
  }
  return rtlRender(ui, { wrapper: Wrapper, ...renderOptions });
}
// override render method
export { render };
