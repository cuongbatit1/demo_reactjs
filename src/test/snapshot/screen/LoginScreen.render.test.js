import { screen } from "@testing-library/react";
import LoginScreen from "../../../ScreenPage/LoginScreen/LoginScreen";
import { render } from "../../../setupTests";

test("test render LoginScreen snapshot", () => {
  const { asFragment } = render(<LoginScreen />);
  expect(asFragment()).toMatchSnapshot();
});
