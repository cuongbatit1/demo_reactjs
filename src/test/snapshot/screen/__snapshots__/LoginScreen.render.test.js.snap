// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test render LoginScreen snapshot 1`] = `
<DocumentFragment>
  <div
    class="css-1jhze1m"
  >
    <div
      class="MuiGrid-root MuiGrid-container MuiGrid-direction-xs-column css-ujljcm-MuiGrid-root"
    >
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-container css-1dgaazi-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-3fd9z3-MuiGrid-root"
          >
            <div
              class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-vlo4ox-MuiPaper-root-MuiCard-root"
            >
              <div
                class="MuiBox-root css-1arlb8v"
              >
                <div
                  class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-919fip-MuiGrid-root"
                >
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
                  >
                    <h6
                      class="MuiTypography-root MuiTypography-subtitle1 css-y4bqle-MuiTypography-root"
                    >
                      Login with email
                    </h6>
                    <div
                      class="MuiFormControl-root MuiTextField-root css-1sqs9ir-MuiFormControl-root-MuiTextField-root"
                      style="width: 100%;"
                    >
                      <label
                        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="true"
                        for=":r0:"
                        id=":r0:-label"
                      >
                        Email
                      </label>
                      <div
                        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiOutlinedInput-input MuiInputBase-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                          id=":r0:"
                          type="email"
                          value="admin@gmail.com"
                        />
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1in441m"
                          >
                            <span>
                              Email
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                    </div>
                    <div
                      class="MuiFormControl-root css-1r7ruti-MuiFormControl-root"
                    >
                      <label
                        class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="true"
                        for="outlined-adornment-password"
                      >
                        Password
                      </label>
                      <div
                        class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                          id="outlined-adornment-password"
                          type="password"
                          value="123456"
                        />
                        <div
                          class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                        >
                          <button
                            aria-label="toggle password visibility"
                            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
                            tabindex="0"
                            type="button"
                          >
                            <svg
                              aria-hidden="true"
                              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                              data-testid="VisibilityIcon"
                              focusable="false"
                              viewBox="0 0 24 24"
                            >
                              <path
                                d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                              />
                            </svg>
                            <span
                              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                            />
                          </button>
                        </div>
                        <fieldset
                          aria-hidden="true"
                          class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                        >
                          <legend
                            class="css-1in441m"
                          >
                            <span>
                              Password
                            </span>
                          </legend>
                        </fieldset>
                      </div>
                      <p
                        class="MuiFormHelperText-root MuiFormHelperText-sizeMedium MuiFormHelperText-contained MuiFormHelperText-filled css-1wc848c-MuiFormHelperText-root"
                      />
                    </div>
                    <div
                      class="MuiGrid-root MuiGrid-container css-pwltme-MuiGrid-root"
                    >
                      <div
                        class="MuiBox-root css-0"
                      >
                        <span
                          class="MuiCheckbox-root MuiCheckbox-colorPrimary MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorPrimary PrivateSwitchBase-root css-12wnr2w-MuiButtonBase-root-MuiCheckbox-root"
                          style="color: black;"
                        >
                          <input
                            aria-label="select all desserts"
                            class="PrivateSwitchBase-input css-1m9pwf3"
                            data-indeterminate="false"
                            type="checkbox"
                          />
                          <svg
                            aria-hidden="true"
                            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                            data-testid="CheckBoxOutlineBlankIcon"
                            focusable="false"
                            viewBox="0 0 24 24"
                          >
                            <path
                              d="M19 5v14H5V5h14m0-2H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                            />
                          </svg>
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </span>
                      </div>
                      <h2
                        class="MuiTypography-root MuiTypography-h6 css-1juivf6-MuiTypography-root"
                      >
                        Remember me
                      </h2>
                    </div>
                    <button
                      class="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-nm0o5x-MuiButtonBase-root-MuiButton-root"
                      tabindex="0"
                      type="button"
                    >
                      Sign in
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
                  >
                    <hr
                      class="MuiDivider-root MuiDivider-fullWidth css-9mgopn-MuiDivider-root"
                    />
                  </div>
                  <div
                    class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
                  >
                    <div
                      class="MuiGrid-root MuiGrid-container MuiGrid-item MuiGrid-direction-xs-column MuiGrid-grid-xs-12 css-clh38r-MuiGrid-root"
                    >
                      <a
                        class="MuiTypography-root MuiTypography-subtitle1 css-93pbn4-MuiTypography-root"
                        href="/register/"
                      >
                        Don't have an account?
                      </a>
                      <a
                        class="MuiTypography-root MuiTypography-subtitle1 css-93pbn4-MuiTypography-root"
                        href="/about/"
                      >
                        Go to About
                      </a>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1lct78s-MuiGrid-root"
      />
    </div>
  </div>
</DocumentFragment>
`;
